workflow/controller/operator.go

```diff
@@ -1460,16 +1461,21 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 			node = lastChildNode
 		} else {
 			// Create a new child node and append it to the retry node.
-			nodeName = fmt.Sprintf("%s(%d)", retryNodeName, len(retryParentNode.Children))
+			childrenLength := strconv.Itoa(len(retryParentNode.Children))
+			nodeName = fmt.Sprintf("%s(%s)", retryNodeName, childrenLength)
 			woc.addChildNode(retryNodeName, nodeName)
 			node = nil
 
+			localParams := make(map[string]string)
 			// Change the `pod.name` variable to the new retry node name
 			if processedTmpl.IsPodType() {
-				processedTmpl, err = common.SubstituteParams(processedTmpl, map[string]string{}, map[string]string{common.LocalVarPodName: woc.wf.NodeID(nodeName)})
-				if err != nil {
-					return woc.initializeNodeOrMarkError(node, nodeName, wfv1.NodeTypeSkipped, templateScope, orgTmpl, opts.boundaryID, err), err
-				}
+				localParams[common.LocalVarPodName] = woc.wf.NodeID(nodeName)
+			}
```

Comment:
From the old code, it seems like we want to perform the parameter substitution only if `processedTmpl.IsPodType()` is true. But with your changes you don't respect this requirement. Is there a reason for this? Otherwise, just make sure that the parameter substitution happens inside the `if` block as was done previously.

Reply:
this happens to support non pod objects like `TemplateTypeSteps` and `TemplateTypeDAG`. If you see the second test it demonstrates this change working with a `TemplateTypeSteps` template which, based on the execution model of Argo, `retryAttempt` must be passed from the step to the execution template.

Reply:
Ah I see, make sense!